src/factor_instances_provider/provider/factor_instances_cache.rs

```diff
@@ -0,0 +1,525 @@
+        network_id: NetworkID,
+    ) -> Result<CachedInstancesWithQuantitiesOutcome> {
+        let target_quantity = originally_requested_quantified_derivation_preset.quantity;
+        let mut pf_instances = IndexMap::<FactorSourceIDFromHash, FactorInstances>::new();
+        let mut pf_pdp_qty_to_derive =
+            IndexMap::<FactorSourceIDFromHash, IndexMap<DerivationPreset, usize>>::new();
+        let mut is_quantity_satisfied_for_all_factor_sources = true;
+
+        for factor_source_id in factor_source_ids {
+            for preset in DerivationPreset::all() {
+                let index_agnostic_path = preset.index_agnostic_path_on_network(network_id);
+                let for_preset = self
+                    .get_mono_factor(factor_source_id, index_agnostic_path)
+                    .unwrap_or_default();
+                let count_in_cache = for_preset.len();
+                if preset == originally_requested_quantified_derivation_preset.derivation_preset {
+                    let satisfies_requested_quantity = count_in_cache >= target_quantity;
+                    if satisfies_requested_quantity {
+                        // The instances in the cache can satisfy the requested quantity
+                        // for this factor source for this derivation preset
+                        pf_instances.append_or_insert_to(
+                            factor_source_id,
+                            // Only take the first `target_quantity` instances
+                            // to be used, the rest are not needed and should
+                            // remain in the cache (later we will call delete on
+                            // all those instances.)
+                            for_preset.split_at(target_quantity).0,
+                        );
+                    } else {
+                        // The instances in the cache cannot satisfy the requested quantity
+                        // we must derive more!
+                        is_quantity_satisfied_for_all_factor_sources = false;
+                        // Since we are deriving more we might as well ensure that the
+                        // cache is filled with `CACHE_FILLING_QUANTITY` **AFTER** the
+                        // requested quantity is satisfied, meaning we will not only
+                        // derive `CACHE_FILLING_QUANTITY - count_in_cache`, instead we
+                        // derive the `target_quantity` as well.
+                        let quantity_to_derive =
+                            CACHE_FILLING_QUANTITY - count_in_cache + target_quantity;
+                        pf_pdp_qty_to_derive.append_or_insert_element_to(
+                            factor_source_id,
+                            (preset, quantity_to_derive),
+                        );
+                        // insert all instances to be used directly
+                        pf_instances.append_or_insert_to(factor_source_id, for_preset.clone());
```

Comment:
This line, 239, we know `for_preset.len()` (`count_in_cache`) is less than `target_quantity`. So we can `append_or_insert` ALL of them, as done on this line. And then we know that since those are less than `target_quantity` we also need to derive more.

How many more? `let quantity_to_derive = CACHE_FILLING_QUANTITY - count_in_cache + target_quantity;` more. 

Both in this scope and the `else` scope below (`preset != originally_requested_derivation_preset`) we derive (at least) `CACHE_FILLING_QUANTITY - count_in_cache` many more. Reason being if e.g. `target_qty` is 28, and `count_in_cache` is 26 (and `CACHE_FILLING_QUANTITY` is 30) we lack 28-26 = 2 to satisfy `target_qty` so we must derive 2 more, but then we would finish with an empty cache, so we want to derive `CACHE_FILLING_QUANTITY` more than that = 2+30. So derive `target_quantity - count_in_cache + CACHE_FILLING_QUANTITY`, which is what we have here: 
`let quantity_to_derive = CACHE_FILLING_QUANTITY - count_in_cache + target_quantity;` (reordered the terms).

In the example with `target_qty = 28` and `count_in_cache = 26` above, we will put the 26 instances into `pf_instances`, we will derive 32 instances. We will later (not here, done in FactorInstancesProvider) append the **first count_in_cache - target_quantity** = 2 instances of the newly derived to the `for_preset` found in cache to get 26+2 = 28 = `target_qty` many instances to use directly (not cached, returned by `for_account_mfa`) and left to cache is 32-2 = 30 = `CACHE_FILLING_QUANTITY` instances to cache. Leaving the cache FULL for this factor source for this preset.

Did I get that right now?

Reply:
I think this is the source of confusion:

For this requested preset we want IN TOTAL to have `CACHE_FILLING_QUANTITY` many instances left in cache after original request is done and we want to return `target_quantity` many instances. and in this case we know `count_in_cache` is less than `target_quantity` `let missing_from_cache_to_satisfy_target = target_quantity - count_in_cache`. So we MUST derive `missing_from_cache_to_satisfy_target` to satisfy original request.

And since we are deriving ANYWAY we might as well also derive `CACHE_FILLING_QUANTITY` so that the cache is full at end of execution.

So we will derive `missing_from_cache_to_satisfy_target + CACHE_FILLING_QUANTITY` but since `missing_from_cache_to_satisfy_target == target_quantity - count_in_cache` => we must derive:
`target_quantity - count_in_cache + CACHE_FILLING_QUANTITY` <==> `CACHE_FILLING_QUANTITY - coun_in_cache + target_quantity`. Exactly what is in the code.

And of those derived we will use `missing_from_cache_to_satisfy_target` many of them DIRECTLY (not cache). and we will cache `CACHE_FILLING_QUANTITY` many of them. 

I think this all adds up. Right?

Reply:
Maybe the confusion comes from the fact that **we would not derive more instances in this function**, rather this function will use as many instances there are + it will report in response how many more need to be derived.

Reply:
Thanks for the detailed explanation!

> In the example with target_qty = 28 and count_in_cache = 26 above, we will put the 26 instances into pf_instances, we will derive 32 instances. We will later (not here, done in FactorInstancesProvider) append the first count_in_cache - target_quantity = 2 instances of the newly derived to the for_preset found in cache to get 26+2 = 28 = target_qty many instances to use directly (not cached, returned by for_account_mfa) and left to cache is 32-2 = 30 = CACHE_FILLING_QUANTITY instances to cache. Leaving the cache FULL for this factor source for this preset.


I believe this answer my question.

So when I looked at this code, I noticed that we only put the 26 instances into `pf_instances`. Based on your explanation, the remaining instances will be appended from` FactorInstancesProvider` [here](https://github.com/radixdlt/sargon-mfa/blob/c6ef7575a6796357c1be513282db4319db921072/src/factor_instances_provider/provider/factor_instances_provider.rs#L140C22-L140C43). Understood now 👍

Reply:
Yeah this is not an easy problem to grasp in an hour.

I've worked on this for a looong time!